lms/services/canvas.py

```diff
@@ -1,85 +1,119 @@
-        "unpublished" in Canvas can only be seen by teachers, not students).
+        :raise CanvasAPIPermissionError: if the user gets a permissions error
+            from the Canvas API when trying to get a public URL for file_id
         """
-        files_in_course = self.api.list_files(course_id)
-
-        for file_ in files_in_course:
-
+        # If there's a previously stored mapping for file_id use that instead.
+        effective_file_id = module_item_configuration.get_canvas_mapped_file_id(file_id)
+
+        try:
+            if check_in_course:
+                self._finder.assert_file_in_course(course_id, effective_file_id)
+            return self.api.public_url(effective_file_id)
+        except (CanvasFileNotFoundInCourse, CanvasAPIPermissionError):
+            # The user can't see the file in the course. This could be because:
+            #
+            # * The course has been copied so the assignment's file_id is from
+            #   the original course (and the user isn't in the original course)
+            # * The assignment's file has been deleted from Canvas
+            # * The file *is* in the course but the user can't see it
+            #   (only instructors can see "unpublished" files in Canvas)
+            #
+            # We'll try to find another copy of the same file that the current
+            # user *can* see in the current course and use that instead.
+            found_file_id = self._finder.find_matching_file_in_course(
+                course_id, effective_file_id
+            )
+
+            if not found_file_id:
+                raise
+
+            # Try again to return a public URL, this time using found_file_id.
+            url = self.api.public_url(found_file_id)
+
+            # Store a mapping so we don't have to re-search next time.
+            module_item_configuration.set_canvas_mapped_file_id(file_id, found_file_id)
+
+            return url
+
+
+class CanvasFileFinder:
+    """A helper for finding file IDs in the Canvas API."""
+
+    def __init__(self, canvas_api, file_service):
+        self._api = canvas_api
+        self._file_service = file_service
+
+    def assert_file_in_course(self, course_id: str, file_id: str) -> bool:
+        """Raise if the current user can't see file_id in course_id."""
+        for file in self._api.list_files(course_id):
             # The Canvas API returns file IDs as ints but the file_id param
             # that this method receives (from our proxy API) is a string.
             # Convert ints to strings so that we can compare them.
-            if str(file_["id"]) == file_id:
-                return True
+            if str(file["id"]) == file_id:
+                return
 
-        return False
+        raise CanvasFileNotFoundInCourse(file_id)
 
     def find_matching_file_in_course(
-        self, course_id: str, file_: models.File
+        self, course_id: str, file_id: str
```

Comment:
```suggestion
        self, course_id, file_id
```

In our chat with Rob I think we said it was best to avoid starting documenting all basic types in our function definitions. I think this could lead to quite a lot of visual noise.

Reply:
Yeah, I just got rid of all these